File: R/tools/list_tools.R
import::here(stringi, 'stri_replace_all_regex')

## Functions

## dict_zip
## fill_missing_keys
## filter_list_for_match
## find_first_match_index
## interleave
## items_in_a_not_b
## list2matrix
## matrix2list
## move_list_items_to_front
## multiple_replacement
## replace_specific_items


#' Convert a matrix into a list of lists
#'
#' Input:
#'      [,1] [,2] [,3] [,4] [,5] [,6]
#' [1,]    1    1    1    2    2    3
#' [2,]    2    3    4    3    4    4
#' 
#' Output: 
#' list(c(1, 2), c(1, 3), c(1, 4), c(2, 4), c(2, 4), c(3, 4))
#' 
collect_matrix_cols <- function(mat) {
    return(split(mat, col(mat)))
}


#' Dictionary
#' 
#' @description Simple dictionary implementation using R environment
#'
#' @export
dict_zip <- function(keys, values) {
    if (!is.list(values)) {
        values <- as.list(values)
    }
    named_list <- setNames(values, keys)
    env <- list2env(named_list)
    return(env)
}


#' Fill Missing Keys
#' 
#' Add missing keys to a named list (ie. dict)
#' 
fill_missing_keys <- function(x, keys, val=NA) {

    missing <- setdiff(keys, names(x))
    x[missing] <- val  # fill
    x <- x[keys]  # sort
    return(x)
}


#' Find matches based on substring
#' 
#' @description Return elements of a list matching a particular substring
#' 
#' @param items list or vector
#' @param patterns a string or collection of strings
#' @return Returns a list or vector with any matching items
#' 
#' @examples
#' filter_list_for_match(c("a_suffix", "b_suffix", "c"), "suffix")
#' 
filter_list_for_match <- function(items, patterns) {
    # filter
    for (i in 1:length(patterns)){
        items <- lapply(items, grep, pattern=patterns[[i]], value=TRUE)
    }
    return (unlist(items[!sapply(items, identical, character(0))]))  # remove character(0)
}


#' Returns the first match given a pattern
#' 
#' @description Thin wrapper around [grep()]
#' 
#' @param items list or vector
#' @param pattern a single regex pattern, can also be an exact string
#' @return Returns the index of the first match
#' 
#' @examples
#' find_first_match_index('_suffix', c('a', 'b_suffix', 'c'))
#' 
find_first_match_index <- function(pattern, items) {
    return (grep(pattern, items)[[1]])
}


#' Interleave
#' 
#' @description
#' 
#' @param a list or vector
#' @param b list or vector
#' @return Returns the interleaved list
#' 
#' @references
#' \href{https://stackoverflow.com/questions/16443260/interleave-lists-in-r}{Stack Overflow}
#' 
#' @examples
#' interleave(c('A', 'B', 'C'), c('a', 'b', 'c'))
#' 
interleave <- function(a, b) {
    idx <- order(c(seq_along(a), seq_along(b)))
    return(unlist(c(a,b))[idx])
}


#' Return items unique to vector a
#' 
#' @description
#' Return all the items found in a and not b. Useful for filtering dataframe columns.
#' 
#' @param a list or vector
#' @param b list or vector
#' @return Returns the filtered list or vector (same type as a)
#' 
#' @references
#' \href{https://stackoverflow.com/questions/10298662/find-elements-not-in-smaller-character-vector-list-but-in-big-list}{Stack Overflow}
#' 
#' @examples
#' items_in_a_not_b(c('a', 'b', 'c', '1', '2'), c('1', '2'))
#' 
items_in_a_not_b <- function(a, b) {
    return((new <- a[which(!a %in% b)]))
}


#' Rearrange a list into a matrix
#' 
#' @description Thin wrapper around matrix, inserts NA
#' 
#' @param z list or vector
#' @param ncol
#' @param byrow
#' @param fill
#' @return 
#' 
#' @examples
#' list2matrix(c('a', 'b', 'c', 'd', 'e'), ncol=3 fill=NA)
#' 
list2matrix <- function(z, ncol=2, byrow=TRUE, fill=NA) {
    remain <- length(z) %% ncol
    if (remain > 0) {
        z <- c(z, rep(fill, ncol-remain))
    }
    mtx <- matrix(z, ncol = ncol, byrow = byrow)
    return(mtx)
}


#' Rearrange a matrix to a named list
#' 
matrix2list <- function(mat) {
    vec <- as.vector(mat)
    names(vec) <- with(
        expand.grid(rownames(mat), colnames(mat)), 
        paste(Var1, Var2, sep = "-")
    )
    return(unlist(vec))
}


#' Rearrange items according to an ordered list
#' 
#' @description Front-loads the ordered list, then adds the remaining items
#' 
#' @param items list or vector
#' @param ordered
#' @return 
#' 
#' @examples
#' move_list_items_to_front(c('a', 'b', 'c', 'd'), c('b', 'c'))
#' 
move_list_items_to_front <- function(items, ordered) {
    return(c(ordered[ordered %in% items],
             items_in_a_not_b(items, ordered)))
}


#' Replaces each item in a list or vector with all replacements
#' 
#' @description
#' Convenience function to perform multiple replacements on a list or dataframe column.
#' Unlike [replace_specific_items()], `multiple_replacement()` can recognize patterns.
#' Note: this is equivalent to [plyr::mapvalues()]
#' 
#' @param items list or vector
#' @param replacements a named list of replacements. uses names to match and values to replace.
#' @return Returns a vector with replaced items
#' 
#' @examples
#' replacements <- c('prefix_' = '', '_suffix' = '')
#' items <- c('prefix_a_suffix', 'prefix_b_suffix')
#' multiple_replacement(items, replacements)
#' 
#' @seealso [replace_specific_items()], [stringi::stri_replace_all_regex()]
#' 
multiple_replacement <- function(items, replace_dict) {

    # for (pattern in names(replace_dict)) {
    #     replacement = replace_dict[[pattern]]
    #     items <- sapply(items, function(x) gsub(pattern, replacement, x))
    # }

    patterns <- names(replace_dict)
    replacements <- sapply(unname(replace_dict), function(x) gsub('\\\\', '$', x))
    
    items <- sapply(items,
        function(x) stri_replace_all_regex(
            x,
            pattern = patterns,
            replacement = replacements,
            vectorize_all = FALSE)
    )

    return (items)
}


#' Replaces each item in a list or vector with all replacements
#' 
#' @description Use this to rename columns
#' 
#' @param items list or vector
#' @param replacements a named list of replacements. uses names to match and values to replace.
#' @return Returns a vector with replaced items
#' 
#' @examples
#' replace_specific_items(c('a', 'b', 'c'), c('a'="A", 'c'="C"))
#' 
#' @seealso [multiple_replacement()]
#' 
replace_specific_items <- function(items, replacer) {
    replace_ids <- which(items %in% intersect(names(replacer), items))
    for (idx in replace_ids) {
        items[idx] <- replacer[items[idx]]
    }
    return(items)
}

File: python/tools/list_tools.py
import re
from typing import Any, List, Dict, Union
from itertools import combinations


def collect_matrix_cols(mat):
    """Convert a matrix into a list of tuples (column-wise)."""
    return [tuple(col) for col in zip(*mat)]


def dict_zip(keys, values):
    """Create a dictionary from keys and values."""
    return dict(zip(keys, values))


def fill_missing_keys(x: dict, keys: list, val=None) -> dict:
    """Add missing keys to a dictionary with a default value."""
    result = x.copy()
    for key in keys:
        if key not in result:
            result[key] = val
    return {k: result.get(k, val) for k in keys}


def filter_list_for_match(items: list, patterns: Union[str, list]) -> list:
    """Return elements of a list matching particular substring(s)."""
    if isinstance(patterns, str):
        patterns = [patterns]

    result = items.copy() if isinstance(items, list) else list(items)
    for pattern in patterns:
        result = [item for item in result if pattern in str(item)]
    return result


def find_first_match_index(pattern: str, items: list) -> int:
    """Return the index of the first match given a pattern."""
    for i, item in enumerate(items):
        if re.search(pattern, str(item)):
            return i
    raise ValueError(f"No match found for pattern: {pattern}")


def interleave(a: list, b: list) -> list:
    """Interleave two lists."""
    result = []
    for i in range(max(len(a), len(b))):
        if i < len(a):
            result.append(a[i])
        if i < len(b):
            result.append(b[i])
    return result


def items_in_a_not_b(a: list, b: list) -> list:
    """Return items in a that are not in b."""
    b_set = set(b)
    return [item for item in a if item not in b_set]


def list2matrix(z: list, ncol: int = 2, byrow: bool = True, fill=None) -> list:
    """Rearrange a list into a matrix (list of lists)."""
    remain = len(z) % ncol
    if remain > 0:
        z = list(z) + [fill] * (ncol - remain)

    if byrow:
        return [z[i:i+ncol] for i in range(0, len(z), ncol)]
    else:
        nrow = len(z) // ncol
        return [[z[j*nrow + i] for j in range(ncol)] for i in range(nrow)]


def matrix2list(mat: list) -> dict:
    """Rearrange a matrix to a named dictionary."""
    result = {}
    for i, row in enumerate(mat):
        for j, val in enumerate(row):
            result[f"{i}-{j}"] = val
    return result


def move_list_items_to_front(items: list, ordered: list) -> list:
    """Rearrange items with ordered items at front."""
    front = [x for x in ordered if x in items]
    back = items_in_a_not_b(items, ordered)
    return front + back


def multiple_replacement(items: Union[str, list], replace_dict: dict) -> Union[str, list]:
    """Replace items using regex patterns from a dictionary."""
    is_string = isinstance(items, str)
    if is_string:
        items = [items]

    result = []
    for item in items:
        text = str(item)
        for pattern, replacement in replace_dict.items():
            text = re.sub(pattern, replacement, text)
        result.append(text)

    return result[0] if is_string else result


def replace_specific_items(items: list, replacer: dict) -> list:
    """Replace specific items in a list using exact matches."""
    return [replacer.get(item, item) for item in items]
